Bluetooth: UBSAN warnings fixes

ivaniushkov opened the pull request:
Several warnings were reported by UBSAN when testing different Bluetooth LE features.
This PR fixes 3 different warnings.

ivaniushkov committed:
Bluetooth: fixing UBSAN warnings related to Codec Configuration

During local testing with UBSAN enabled, warning was reported:
bluetooth/host/iso.c:237:2: runtime error: null pointer passed
as argument 2, which is declared to never be null

It turned out that when datapath doesn't contain
codec information, cc_len is 0 and cc is NULL

In order to avoid UB,
now we call memcpy only when cp->codec_config_len > 0

Signed-off-by: Ivan Iushkov <ivan.iushkov@nordicsemi.no>

ivaniushkov committed:
Bluetooth: fixing null-pointer dereference in l2cap channel destroyer

During local testing with UBSAN enabled, warning was reported:
bluetooth/host/l2cap.c:980:25: runtime error: member access
within null pointer of type 'struct k_work_q'

It turned out that le_chan->rtx_work.queue can be NULL.
Since null-pointer dereference is a UB, additional check
was added to ensure we don't access
`le_chan->rtx_work.queue->thread` when
`le_chan->rtx_work.queue == NULL`

The same changes applied to l2cap_br.c

Signed-off-by: Ivan Iushkov <ivan.iushkov@nordicsemi.no>

Thalley commented on subsys/bluetooth/host/adv.c:
```diff
@@ -2064,8 +2064,12 @@ void bt_hci_le_per_adv_response_report(struct net_buf *buf)
+		if (response->cte_type == BT_HCI_LE_NO_CTE) {
+			info.cte_type = BT_DF_CTE_TYPE_NONE;
+		} else {
+			info.cte_type = BIT(response->cte_type);
```
This assumes that the `BT_DF_CTE_TYPE_*` values are defined in a specific way. We generally do not make such assumptions which is why we have functions like

```c
static uint8_t get_hci_cte_type(enum bt_df_cte_type type)
{
	switch (type) {
	case BT_DF_CTE_TYPE_AOA:
		return BT_HCI_LE_AOA_CTE;
	case BT_DF_CTE_TYPE_AOD_1US:
		return BT_HCI_LE_AOD_CTE_1US;
	case BT_DF_CTE_TYPE_AOD_2US:
		return BT_HCI_LE_AOD_CTE_2US;
	default:
		LOG_ERR("Wrong CTE type");
		return BT_HCI_LE_NO_CTE;
	}
}
```

Suggest to create a function that converts HCI types to `BT_DF_CTE_TYPE_*` to make it future proof.

ivaniushkov replied:
I think it will require adding this function to direction.c/direction.h and adding direction.h as a dependency to adv.c.
I believe it is a redundant dependency, so I would suggest to keep the implementation as is for now

Thalley replied:
It does not need to be a public function

ivaniushkov replied:
I added an inline function to direction.h
Please let me know if you would prefer other implementation

Thalley commented on subsys/bluetooth/host/scan.c:
```diff
@@ -825,9 +825,13 @@ static void bt_hci_le_per_adv_report_common(struct net_buf *buf)
+	if (evt->cte_type == BT_HCI_LE_NO_CTE) {
+		info.cte_type = BT_DF_CTE_TYPE_NONE;
+	} else {
+		info.cte_type = BIT(evt->cte_type);
```
Ditto

jhedberg commented on subsys/bluetooth/host/iso.c:
```diff
@@ -235,8 +235,9 @@ static int hci_le_setup_iso_data_path(const struct bt_conn *iso, uint8_t dir,
 	cc = net_buf_add(buf, cp->codec_config_len);
-	memcpy(cc, path->cc, cp->codec_config_len);
-
+	if (cp->codec_config_len) {
+		memcpy(cc, path->cc, cp->codec_config_len);
```
Would it perhaps be better to use `path->cc_len` instead of `cp->codec_config_len` here? `cp` is a packed struct, so it's possible that accessing its members generates less efficient code than accessing `path->cc_len`.

ivaniushkov replied:
Yes, you are right, I changed it to `if (path->cc_len)`

Thalley commented on include/zephyr/bluetooth/direction.h:
```diff
@@ -359,4 +359,26 @@ int bt_df_conn_cte_rsp_enable(struct bt_conn *conn);
+ *
+ * @return CTE type (@ref bt_df_cte_type).
+ */
+static inline int bt_df_get_bt_df_cte_type(uint8_t hci_cte_type)
+{
+	switch (hci_cte_type) {
+	case BT_HCI_LE_AOA_CTE:
+		return BT_DF_CTE_TYPE_AOA;
+	case BT_HCI_LE_AOD_CTE_1US:
+		return BT_DF_CTE_TYPE_AOD_1US;
+	case BT_HCI_LE_AOD_CTE_2US:
+		return BT_DF_CTE_TYPE_AOD_2US;
+	case BT_HCI_LE_NO_CTE:
+		return BT_DF_CTE_TYPE_NONE;
+	default:
+		return BT_DF_CTE_TYPE_NONE;
+	}
+}
```
Given that it's unlikely that an application would use this, I would perhaps put it in `direction_internal.h` instead.

Thalley replied:
Hmm, actually, since it's also used in by other files (such as scan.c), perhaps place it next to `bt_get_phy` which does the same thing, but for PHYs. 

For something else (but somewhat related), we seemingly have both `BT_DF_CTE_TYPE_AOA` and `BT_GAP_CTE_AOA` (and duplicates of the others as well). 

Doesn't need to be fixed in this PR, but ideally we should only define those values one place.

ivaniushkov replied:
Thank you, I moved it to direction_internal.h

ivaniushkov replied:
I moved a conversion function to hci_core next to bt_get_phy()

ivaniushkov committed:
Bluetooth: Fixing UBSAN warning in CTE field parsing in adv.c/scan.c

during local testling, UBSAN reported the following warnings:
- bluetooth/host/adv.c:2067:19: runtime error: shift exponent
255 is too large for 32-bit type 'long unsigned int'
- bluetooth/host/scan.c:828:18: runtime error: shift exponent
255 is too large for 32-bit type 'long unsigned int'

It turned out that we can't use BIT() macro directly on
bt_hci_evt_le_per_advertising_report::cte_type field.
According to Core Spec, `cte_type = 0xFF` corresponds
to `No contstant tone extension`.

Added separate function to convert CTE bit field from
HCI format to bt_df_cte_type

Signed-off-by: Ivan Iushkov <ivan.iushkov@nordicsemi.no>

Thalley commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -424,6 +424,22 @@ uint8_t bt_get_phy(uint8_t hci_phy)
 	}
 }
 
+int bt_get_bt_df_cte_type(uint8_t hci_cte_type)
```
```suggestion
int bt_get_df_cte_type(uint8_t hci_cte_type)
```

ivaniushkov replied:
Thanks, I renamed it

ivaniushkov commented on subsys/bluetooth/host/iso.c:
```diff
@@ -235,8 +235,9 @@ static int hci_le_setup_iso_data_path(const struct bt_conn *iso, uint8_t dir,
-	memcpy(cc, path->cc, cp->codec_config_len);
-
+	if (path->cc_len) {
+		memcpy(cc, path->cc, cp->codec_config_len);
```
Thanks, I also changed `cc = net_buf_add(buf, cp->codec_config_len)` to `cc = net_buf_add(buf, path->cc_len)` for consistency
